File: packages/workspace_flutter/ios/Classes/Internal/WorkspaceOperationEngine.swift
import Foundation

/**
 Owns operation registration, background scheduling, cleanup barriers, and
 generation-fenced main-thread settlement. Channel handlers provide protocol
 validation and provider work, then translate only the terminal wire payload.
 */
internal final class WorkspaceOperationEngine {
  private let registry = WorkspaceOperationRegistry()
  private let stateLock = NSLock()
  private var generation = 0
  private var detached = false

  func run(
    operationId: String,
    workspaceId: String,
    remainingMillis: Int64,
    settle: @escaping (WorkspaceNativeOperation, Result<Any, Error>) -> Void,
    body: @escaping (WorkspaceNativeOperation, TimeInterval) throws -> Any
  ) -> Bool {
    stateLock.lock()
    let currentGeneration = generation
    let isDetached = detached
    stateLock.unlock()
    guard !isDetached,
          let operation = registry.register(
            operationId: operationId,
            workspaceId: workspaceId,
            engineGeneration: currentGeneration
          ) else { return false }

    let deadline = ProcessInfo.processInfo.systemUptime + Double(remainingMillis) / 1000.0
    DispatchQueue.global(qos: .userInitiated).async {
      let bodyTerminal: Result<Any, Error>
      do {
        bodyTerminal = .success(try body(operation, deadline))
      } catch {
        bodyTerminal = .failure(error)
      }
      let terminal = operation.captureBody(bodyTerminal)
      DispatchQueue.main.async {
        defer { self.registry.remove(operationId) }
        guard self.canSettle(operation) else { return }
        settle(operation, terminal)
      }
    }
    return true
  }

  func cancel(_ operationId: String) { registry.cancel(operationId) }

  func closeWorkspace(_ workspaceId: String) { registry.closeWorkspace(workspaceId) }

  func isWorkspaceClosing(_ workspaceId: String) -> Bool {
    registry.isWorkspaceClosing(workspaceId)
  }

  func afterWorkspaceCleanup(_ workspaceId: String, _ cleanup: @escaping () -> Void) {
    DispatchQueue.global(qos: .userInitiated).async {
      self.registry.awaitWorkspaceCleanup(workspaceId)
      cleanup()
    }
  }

  func detach() {
    stateLock.lock()
    detached = true
    generation += 1
    stateLock.unlock()
    registry.cancelAll()
  }

  private func canSettle(_ operation: WorkspaceNativeOperation) -> Bool {
    stateLock.lock()
    defer { stateLock.unlock() }
    return !detached && operation.engineGeneration == generation
  }
}

File: packages/workspace_flutter/android/src/main/kotlin/com/mattsp1290/workspace_flutter/internal/WorkspaceOperationEngine.kt
package com.mattsp1290.workspace_flutter.internal

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import io.flutter.plugin.common.MethodChannel
import java.io.FileNotFoundException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

/** Typed failures deliberately contain only wire-safe error codes. */
internal open class WorkspaceOperationFailure(val code: String) : RuntimeException()

/**
 * Owns native worker scheduling, once-only result settlement, cancellation,
 * and cleanup barriers. Channel handlers provide validation and provider work
 * only; they never own a worker or an active operation directly.
 */
internal class WorkspaceOperationEngine(
  private val main: Handler = Handler(Looper.getMainLooper()),
) {
  private var workers: ExecutorService = Executors.newCachedThreadPool()
  private val operations = WorkspaceOperationRegistry()
  private val detached = AtomicBoolean(false)
  private val generation = AtomicLong(0)

  fun attach() {
    generation.incrementAndGet()
    detached.set(false)
    if (workers.isShutdown) workers = Executors.newCachedThreadPool()
  }

  fun detach() {
    operations.cancelAll()
    detached.set(true)
    generation.incrementAndGet()
    workers.shutdown()
  }

  fun run(
    operationId: String,
    workspaceId: String,
    remainingMillis: Long,
    result: MethodChannel.Result,
    body: (WorkspaceNativeOperation, Long) -> Any,
  ): Boolean {
    if (detached.get()) return false
    val operation = operations.register(operationId, workspaceId, generation.get()) ?: return false
    if (detached.get() || operation.engineGeneration != generation.get()) {
      operations.remove(operationId, operation)
      return false
    }
    val deadline = SystemClock.elapsedRealtime() + remainingMillis
    try {
      workers.execute {
        try {
          finish(operationId, operation, result, value = body(operation, deadline))
        } catch (error: Throwable) {
          val code = when {
            operation.isCancelled() -> "cancelled"
            error is WorkspaceOperationFailure -> error.code
            error is AndroidProviderUnavailable -> "unavailable"
            error is FileNotFoundException -> "notFound"
            error is SecurityException -> "permissionLost"
            else -> "providerFailure"
          }
          finish(operationId, operation, result, errorCode = code)
        }
      }
    } catch (_: RejectedExecutionException) {
      operations.remove(operationId, operation)
      return false
    }
    return true
  }

  fun cancel(operationId: String) = operations.cancel(operationId)

  fun closeWorkspace(workspaceId: String) = operations.closeWorkspace(workspaceId)

  fun isWorkspaceClosing(workspaceId: String): Boolean = operations.isWorkspaceClosing(workspaceId)

  fun afterWorkspaceCleanup(workspaceId: String, cleanup: () -> Unit) {
    workers.execute {
      operations.awaitWorkspaceCleanup(workspaceId)
      cleanup()
    }
  }

  fun postToMain(callback: () -> Unit) = main.post(callback)

  private fun finish(
    operationId: String,
    operation: WorkspaceNativeOperation,
    result: MethodChannel.Result,
    value: Any? = null,
    errorCode: String? = null,
  ) {
    if (!operation.settled.compareAndSet(false, true)) return
    if (detached.get() || operation.engineGeneration != generation.get()) {
      operations.remove(operationId, operation)
      return
    }
    main.post {
      try {
        if (detached.get() || operation.engineGeneration != generation.get()) return@post
        when {
          operation.isCancelled() -> result.error("cancelled", null, null)
          errorCode != null -> result.error(errorCode, null, null)
          else -> result.success(value)
        }
      } finally {
        operations.remove(operationId, operation)
      }
    }
  }
}
